### PR TITLE
Add realtime provisioning rule to recommendation engine

### PR DESCRIPTION
**Pull request by sajjad-moradi**
## Description
In this PR, a rule is added to recommendation engine to provide some recommendations on realtime provisioning. The rule utilizes MemoryEstimator. Memory estimator uses data characteristics - introduced in #6546 - to be able to generate a segment on the fly and then use its allocated memory to estimate the amount memory needed for the real time table.

## Testing Done
Integration test with `RecommenderDriver`

The json representation of the recommendations for the data characteristics provided in the test looks like the following:

image

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/RulesToExecute.java [hunk 1]
```diff
@@ -134,15 +143,20 @@ public boolean isRecommendBloomFilter() {
...
   public enum Rule {
-    FlagQueryRule, 
+    FlagQueryRule,
     KafkaPartitionRule,
     InvertedSortedIndexJointRule,
     NoDictionaryOnHeapDictionaryJointRule, // NoDictionaryOnHeapDictionaryJointRule must go after InvertedSortedIndexJointRule since we do not recommend NoDictionary on cols with indices
     VariedLengthDictionaryRule, // VariedLengthDictionaryRule must go after NoDictionaryOnHeapDictionaryJointRule  since we do not recommend dictionary on NoDictionary cols
     PinotTablePartitionRule, // PinotTablePartitionRule must go after KafkaPartitionRule to recommend realtime partitions, after NoDictionaryOnHeapDictionaryJointRule to correctly calculate record size
     BloomFilterRule,
+    RealtimeProvisioningRule
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/RulesToExecute.java:160]: Add comments here please explaining the new rule.
What would be better is to move these comments as javadoc for the enum and describe each rule in a bullet list. Will look much cleaner IMO.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+import org.apache.pinot.controller.recommender.rules.io.params.RealtimeProvisioningRuleParams;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.DataSizeUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+
+/**
+ * This rule recommends gives some recommendations useful for provisioning real time tables. Specifically it provides
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:48]: (nit) change `this rule recommends gives some recommendations` to `this rule gives some recommendations`

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/IndexConfig.java [hunk 3]
```diff
@@ -80,6 +80,10 @@ public boolean isSortedColumnOverwritten() {
         return _isSortedColumnOverwritten;
     }
 
+    public void setSortedColumnOverwritten(boolean sortedColumnOverwritten) {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/IndexConfig.java:83]: This should not be needed. Already covered in 

```
@JsonSetter(nulls = Nulls.SKIP)
    public void setSortedColumn(String sortedColumn) {
        this._sortedColumn = sortedColumn;
        this._isSortedColumnOverwritten = true;
    }
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/PartitionConfig.java [hunk 4]
```diff
@@ -83,6 +83,18 @@ public void setNumKafkaPartitions(int numKafkaPartitions) {
     this._numKafkaPartitions = numKafkaPartitions;
   }
 
+  public void setNumPartitionsOfflineOverwritten(boolean numPartitionsOfflineOverwritten) {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/PartitionConfig.java:86]: Not sure why this change is needed ?

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  "queriesWithWeights":{
+    "select i from tableName where b in (2,4) and ((a in (1,2,3) and e = 4) or c = 7) and d in ('#VALUES', 23) and t > 500": 1,
+    "select j from tableName where (a=3 and (h = 5 or f >34) and REGEXP_LIKE(i, 'as*')) or ((f = 3  or j in ('#VALUES', 4)) and REGEXP_LIKE(d, 'fl*'))": 2,
+    "select f from tableName where (a=0 or (b=1 and (e in ('#VALUES',2) or c=7))) and TEXT_MATCH(d, 'dasd') and MAX(MAX(h,i),j)=4 and t<3": 4,
+    "select f from tableName where (a=0 and b=1) or c=7 or (d = 7 and e =1)": 2,
+    "select f from tableName where t between 1 and 1000": 2
+  },
+  "qps": 150,
+  "numMessagesPerSecInKafkaTopic":1000,
+  "numRecordsPerPush":10000,
```
Comment (siddharthteotia) [inline pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:138]: I know for unit/integration tests, we may not want to use a higher number of records. But locally, have you tested with higher number of records. IIRC, as part of previous PR, you mentioned going to > = million records ? Just confirming

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java [hunk 6]
```diff
@@ -41,20 +43,20 @@
...
-      assertEquals(extract(metadata, "column.colStringMV.cardinality = (\\d+)"), "350");
-      assertEquals(extract(metadata, "column.colBytes.cardinality = (\\d+)"), "400");
-      assertEquals(extract(metadata, "column.colLong.cardinality = (\\d+)"), "500");
-      assertEquals(extract(metadata, "column.colLongMV.cardinality = (\\d+)"), "550");
-      assertEquals(extract(metadata, "column.colDouble.cardinality = (\\d+)"), "600");
-      assertEquals(extract(metadata, "column.colDoubleMV.cardinality = (\\d+)"), "650");
-      assertEquals(extract(metadata, "column.colDoubleMetric.cardinality = (\\d+)"), "700");
-      assertEquals(extract(metadata, "column.colFloatMetric.cardinality = (\\d+)"), "800");
-      assertEquals(extract(metadata, "column.colTime.cardinality = (\\d+)"), "900");
+      assertEquals(extract(metadata, "column.colInt.cardinality = (\\d+)"), "10");
```
Comment (siddharthteotia) [inline pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java:46]: This test was added as part of the previous PR right? Why has the result changed due to the integration work done in this PR?

File: pinot-controller/src/test/resources/memory_estimation/schema-with-metadata.json [hunk 7]
```diff
@@ -4,90 +4,90 @@
...
       "name": "colInt",
-      "cardinality": 100
+      "cardinality": 10
     },
     {
       "dataType": "INT",
       "name": "colIntMV",
       "singleValueField": false,
-      "cardinality": 150,
+      "cardinality": 15,
```
Comment (siddharthteotia) [inline pinot-controller/src/test/resources/memory_estimation/schema-with-metadata.json:13]: Ok, so looks like this is the reason why the unit test results changed? Why do we need to alter the cardinality etc in this characteristics file ?

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/params/RecommenderConstants.java [hunk 8]
```diff
@@ -80,6 +81,13 @@
     public static final String ERROR_INVALID_QUERY = "Error: query not able to parse, skipped";
   }
 
+  public static class RealtimeProvisioningRule {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/params/RecommenderConstants.java:84]: Javadoc please. I see classes for other rules here don't have javadocs. We can probably add those docs later. But for new code, let's document as much as possible.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  public RealtimeProvisioningRule(InputManager input, ConfigManager output) {
+    super(input, output);
+    _params = input.getRealtimeProvisioningRuleParams();
+  }
+
+  @Override
+  public void run()
+      throws InvalidInputException {
+
+    if (_params == null) {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:68]: So these params will either be default ones you defined in RecommenderConstants or if the user override them through overwritten configs right?

How do we make sure that regardless of whether this is overwritten or default, this rule should be executed only if the table type is HYBRID or REALTIME ?

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    File statsFile = memoryEstimator.initializeStatsHistory();
+    runAndRethrowIOException(() -> memoryEstimator
+        .estimateMemoryUsed(statsFile, numHosts, numHours, totalConsumingPartitions,
+            _params.getRealtimeTableRetentionHours()));
+
+    // extract recommendations
+    extractResults(memoryEstimator, numHosts, numHours, _output.getRealtimeProvisioningRecommendations());
+  }
+
+  private TableConfig createTableConfig(IndexConfig indexConfig, Schema schema) {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:95]: I remember discussing this offline. 
Our internal integration with this tool will display all recommendations (indexing, partitioning, RT) once the user invokes it through the UI.
However, this work for recommending RT configs will first generate the segment right. To generate the segment, RT provisioning tool / MemoryEstimator will require to know the indexes. How are we taking care of this here ?

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/RulesToExecute.java (hunk 1 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/RulesToExecute.java:160]: Each Rule class has its own javadoc explaining what they do. The comments here are for the order of the rules. Good that you pointed this out, I need to add a comment here explaining that this rule needs to be executed as the last rule as it uses the output of the other rules as its input.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java (hunk 2 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:48]: That's the result of last minute refactoring :)
Will do.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/IndexConfig.java (hunk 3 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/IndexConfig.java:83]: Usually deserializers require standard setters to be able to deserialize to objects. We didn't need this before, because `RecommenderDriver.run(input)` returns a string and that's only used in the endpoint. In tests, the output of each rule is begin tested. Now for the RT.Prov.Rule, the new test verify against the output of the run function - which is more comprehensive - and that requires deserialization of the output string.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/PartitionConfig.java (hunk 4 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/configs/PartitionConfig.java:86]: Same as above comment.

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java [hunk 11]
```diff
@@ -242,13 +244,22 @@ public boolean execute()
 
     long maxUsableHostMemBytes = DataSizeUtils.toBytes(_maxUsableHostMemory);
 
-    if (_numRows == 0) {
-      _numRows = DEFAULT_NUMBER_OF_ROWS;
+    MemoryEstimator memoryEstimator;
+    if (segmentProvided) {
```
[the remark this answers is not in the document]
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java:248]: Will do.

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json (hunk 5 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:138]: Yes, I did test it on my mac with 10M rows with high cardinality on different columns - in order of millions - and no issue were observed.

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java (hunk 6 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java:46]: The values here are generated randomly and probability of creating distinct number of values as specified in cardinality metadata gets closer to one if we increase the number of rows and decrease the cardinality. Previous values in the tests led to flaky results. Here I'm just making sure that the test essentially never fails.

File: pinot-controller/src/test/resources/memory_estimation/schema-with-metadata.json (hunk 7 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/resources/memory_estimation/schema-with-metadata.json:13]: Refer to the above comment.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/params/RecommenderConstants.java (hunk 8 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/io/params/RecommenderConstants.java:84]: Here the constants are defined for RealtimeProvisiningRule. I'm not sure adding comments makes it more clear than that. Some experts believe that we should not put comments as much as we can and instead use descriptive names unless there's no way around it. They believe that in code refactoring comments don't get proper attention and easily become outdated.
Anyways, if you still think it's better to write something there, I can simply add it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java (hunk 10 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:95]: We're using the output of other rules on the fly to create table config which is the input RT.Prov.Rule.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java (hunk 9 above)
Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:68]: 1) If you look at `RealtimeProvisioningRuleParams`, two properties don't have default values and user must specify them. For the other four, if user doesn't specify them, the default values on RecommenderConstants will be picked.
2) Since we generate table config on the fly, there's no table type in the input parameters. I tried to emphasize REALTIME by spelling it out in the name of the field in InputManger - `realtimeProvisioningRuleParams` - and also one of the properties of this param pojo is `realtimeTableRetentionHours`.

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java (hunk 6 above)
Comment (siddharthteotia) [reply pinot-controller/src/test/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimatorTest.java:46]: As discussed offline, this is a good way to make test deterministic. Smaller cardinality might ensure that random generator will generate as many unique values as we want but it is still not deterministic. Secondly, smaller cardinality is not very meaningful for the test. So let's fix the generation code to ensure it generates as many unique values as we want in a deterministic style.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    if (_params == null) {
+      // no realtime provisioning params provided; skip
+      return;
+    }
+
+    // prepare input to memory estimator
+    TableConfig tableConfig = createTableConfig(_output.getIndexConfig(), _input.getSchema());
+    long maxUsableHostMemoryByte = DataSizeUtils.toBytes(_params.getMaxUsableHostMemory());
+    int totalConsumingPartitions = _params.getNumPartitions() * _params.getNumReplicas();
+    int ingestionRate = (int) _input.getNumMessagesPerSecInKafkaTopic() / _params.getNumPartitions();
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:77]: (nit) rename this to ingestionRatePerPartition.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimator.java [hunk 13]
```diff
@@ -142,10 +141,10 @@ public MemoryEstimator(TableConfig tableConfig, File sampleCompletedSegment, int
   /**
    * Constructor used for processing the given data characteristics (instead of completed segment)
    */
-  public MemoryEstimator(TableConfig tableConfig, File schemaWithMetadataFile, int numberOfRows, int ingestionRate,
-      long maxUsableHostMemory, int tableRetentionHours) {
+  public MemoryEstimator(TableConfig tableConfig, Schema schema, SchemaWithMetaData schemaWithMetadata,
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimator.java:144]: (nit) please rename `ingestionRate` to `ingestionRatePerPartition`

File: pinot-controller/src/test/resources/memory_estimation/schema-with-metadata__dateTimeFieldSpec.json [hunk 14]
```diff
@@ -5,34 +5,34 @@
       "dataType": "INT",
       "name": "colInt",
       "singleValueField": false,
-      "cardinality": 500,
+      "cardinality": 50,
```
Comment (siddharthteotia) [inline pinot-controller/src/test/resources/memory_estimation/schema-with-metadata__dateTimeFieldSpec.json:8]: Let's not change these as discussed earlier. Let's make data generation deterministic

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    int[] numHosts = _params.getNumHosts();
+    int[] numHours = _params.getNumHours();
+
+    // run memory estimator
+    MemoryEstimator memoryEstimator =
+        new MemoryEstimator(tableConfig, _input.getSchema(), _input.getSchemaWithMetadata(),
+            (int) _input.getNumRecordsPerPush(), ingestionRate, maxUsableHostMemoryByte,
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java:84]: Please consider adding a TODO here to indicate that we may not want to use numRecordsPerPush as the numRows for the completed segment we are going to generate. A more fine-grained number is needed which we need to figure out how to capture.

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/TestConfigEngine.java [hunk 16]
```diff
@@ -331,23 +340,43 @@ void testFixedLenBitset() {
...
+    String input = readInputToStr("recommenderInput/RealtimeProvisioningInput.json");
+    String output = RecommenderDriver.run(input);
+    ConfigManager configManager = objectMapper.readValue(output, ConfigManager.class);
+    Map<String, Map<String, String>> recommendations = configManager.getRealtimeProvisioningRecommendations();
+    assertRealtimeProvisioningRecommendation(recommendations.get(OPTIMAL_SEGMENT_SIZE));
+    assertRealtimeProvisioningRecommendation(recommendations.get(CONSUMING_MEMORY_PER_HOST));
+    assertRealtimeProvisioningRecommendation(recommendations.get(TOTAL_MEMORY_USED_PER_HOST));
+  }
+
+  private void assertRealtimeProvisioningRecommendation(Map<String, String> matrix) {
```
Comment (siddharthteotia) [inline pinot-controller/src/test/java/org/apache/pinot/controller/recommender/TestConfigEngine.java:374]: This test is not very meaningful since it is not comparing/asserting against the actual output. It is essentially checking the format of the output. Why don't we compare the output agains the pre-buil 3 maps posted in the PR description ? Looks like that is indeed the output of the test.

File: pinot-controller/src/test/resources/memory_estimation/schema-with-metadata__dateTimeFieldSpec.json (hunk 14 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/resources/memory_estimation/schema-with-metadata__dateTimeFieldSpec.json]: Done.

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/TestConfigEngine.java (hunk 16 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/java/org/apache/pinot/controller/recommender/TestConfigEngine.java:374]: This test here makes sure that integration of recommendation engine with MemoryEstimator works by checking the size of the output based on the given parameters.

**Review by siddharthteotia — APPROVED** [review]
LGTM. Thanks for working on this

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+        "averageLength" : 25
+      },
+      {
+        "name": "p",
+        "dataType": "DOUBLE",
+        "cardinality":1000,
+        "numValuesPerEntry":1
+      }
+    ],
+    "timeFieldSpec": {
```
Comment (mcvsubbu) [inline pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:119]: We should do all testing from now on with DateTimeFieldSpec, since TimeFieldSpec is deprecated.

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+      "incomingGranularitySpec": {
+        "dataType": "INT",
+        "name": "t",
+        "timeType": "DAYS",
+        "cardinality":1000,
```
Comment (mcvsubbu) [inline pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:124]: Cardinality is important only on a per-segment basis, right? I am not sure if we are considering that. For example, in an additive table with daily push, the cardinalty of a column with type DAYS is likely 1 or 2 at best.

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json (hunk 17 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:119]: Sid and I actually talked about this and there will be a task for ensuring support for both types of time columns.

File: pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json (hunk 18 above)
Comment (sajjad-moradi) [reply pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:124]: That's correct. Cardinality is only used in per-segment basis as the memory allocated in each segment has correlation with cardinality in that segment and not other segments.

Comment (mcvsubbu) [reply pinot-controller/src/test/resources/recommenderInput/RealtimeProvisioningInput.json:124]: In that case, we should document it clearly in the place where this is obtained (from the table creator).
